Kernel/Arch/aarch64/Processor.cpp

```diff
@@ -112,9 +113,94 @@ void Processor::initialize_context_switching(Thread& initial_thread)
 
 void Processor::switch_context(Thread*& from_thread, Thread*& to_thread)
 {
-    (void)from_thread;
-    (void)to_thread;
-    TODO_AARCH64();
+    VERIFY(!m_in_irq);
+    VERIFY(m_in_critical == 1);
+
+    dbgln_if(CONTEXT_SWITCH_DEBUG, "switch_context --> switching out of: {} {}", VirtualAddress(from_thread), *from_thread);
+
+    // m_in_critical is restored in enter_thread_context
+    from_thread->save_critical(m_in_critical);
+
+    // clang-format off
+    asm volatile(
+        "sub sp, sp, #248 \n"
+        "stp x0, x1,     [sp, #(0 * 0)] \n"
+        "stp x2, x3,     [sp, #(2 * 8)] \n"
+        "stp x4, x5,     [sp, #(4 * 8)] \n"
+        "stp x6, x7,     [sp, #(6 * 8)] \n"
+        "stp x8, x9,     [sp, #(8 * 8)] \n"
+        "stp x10, x11,   [sp, #(10 * 8)] \n"
+        "stp x12, x13,   [sp, #(12 * 8)] \n"
+        "stp x14, x15,   [sp, #(14 * 8)] \n"
+        "stp x16, x17,   [sp, #(16 * 8)] \n"
+        "stp x18, x19,   [sp, #(18 * 8)] \n"
+        "stp x20, x21,   [sp, #(20 * 8)] \n"
+        "stp x22, x23,   [sp, #(22 * 8)] \n"
+        "stp x24, x25,   [sp, #(24 * 8)] \n"
+        "stp x26, x27,   [sp, #(26 * 8)] \n"
+        "stp x28, x29,   [sp, #(28 * 8)] \n"
+        "str x30,        [sp, #(30 * 8)] \n"
+        "mov x0, sp \n"
+        "str x0, %[from_sp] \n"
+        "ldr x0, =1f \n"
+        "str x0, %[from_ip] \n"
+
+        "ldr x0, %[to_sp] \n"
+        "mov sp, x0 \n"
+
+        "sub sp, sp, 24 \n"
+        "ldr x0, %[from_thread] \n"
+        "ldr x1, %[to_thread] \n"
+        "ldr x2, %[to_ip] \n"
+        "str x0, [sp, #0] \n"
+        "str x1, [sp, #8] \n"
+        "str x2, [sp, #16] \n"
+
+        "bl enter_thread_context \n"
+        "ldr x0, [sp, #16]\n"
+        "br x0 \n"
+
+        "1: \n"
```

Comment:
Where is this label used? (Sorry, blind, presumably.)

Reply:
Looks like it's used as a "following" label up above to save the place we''re going to return to after entering the thread context (?)
```
        "str x30,        [sp, #(30 * 8)] \n"
        "mov x0, sp \n"
        "str x0, %[from_sp] \n"
*       "ldr x0, =1f \n"
        "str x0, %[from_ip] \n"
```

Reply:
Yes, as Andrew pointed out this label is used just above using `=1f`.

Right, the address of the `1:` label is stored in the `from_ip` field, such that the thread that gets suspended (or the running thread before the context switch), will resume execution at the `1:` label when it gets scheduled the next time.

This code was very tricky to figure out, and to get right, so I could try to draw a diagram to explain how it all works.